### PR TITLE
Loss functions

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
Added:
-  a number of the most common loss functions.
- one-hot encoding
- Embedding module

Moved the loss functions to the 'Functions' static class, since there's no 'nn.losses' in PyTorch.

### CODE REVIEW

**Comment by dsyme** [comment]
This looks good so far, I'll review again when out of draft

File: src/TorchSharp/NN/Losses.cs [hunk 1]
```diff
@@ -12,21 +12,46 @@ namespace TorchSharp.NN
...
         [DllImport ("LibTorchSharp")]
-        private static extern IntPtr THSNN_binary_cross_entropy (IntPtr srct, IntPtr trgt, IntPtr wgt, long reduction);
+        private static extern IntPtr THSNN_cross_entropy(IntPtr srct, IntPtr trgt, IntPtr wgt, long ignore_index, bool hasII, long reduction);
 
-        public static Loss binary_cross_entropy (TorchTensor? weigths = null, Reduction reduction = Reduction.Mean)
+        public static Loss cross_entropy_loss(TorchTensor? weigths = null, long? ignore_index = null, Reduction reduction = Reduction.Mean)
         {
             return (TorchTensor src, TorchTensor target) => {
-                var res = THSNN_binary_cross_entropy (src.Handle, target.Handle, weigths?.Handle ?? IntPtr.Zero, (long)reduction);
+                var ii = ignore_index.HasValue ? ignore_index.Value : -100;
```
Comment (dsyme) [inline src/TorchSharp/NN/Losses.cs:24]: where's the -100 come from?

**Review by dsyme — APPROVED** [review]

File: src/TorchSharp/NN/Losses.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Losses.cs:24]: This is the default value in the Torch C++ code for this option. Since we do all the defaults on the C# side of the interface, I thought I should use the same.
